include/gsl/pointers

```diff
@@ -113,6 +117,11 @@ private:
     T ptr_;
 };
 
+#if defined GSL_DEDUCTION_GUIDES_SUPPORTED
+template <typename T>
```

Comment:
I like that this elegant little guide that makes deduction work nicely. However, does it point to a failing in our existing set of constructors for `not_null`? Do we just need to have a constructor from `T&&` and then the guide would be unnecessary?

Reply:
> Do we just need to have a constructor from `T&&` and then the guide would be unnecessary?

If there were a `T&&` constructor, the implicit deduction guides would deduce from *rvalue* arguments  - but not *lvalue* arguments. A plain `T` constructor would be a better choice.

Reply:
Makes sense. I was just parrotting the `U&&` one without thinking, `T` makes much more sense. And it shouldn't compete with the `U&&` constructor?

Reply:
Maybe. Will check!

Reply:
> And it shouldn't compete with the `U&&` constructor?

It will compete, and it will win by virtue of being a non-template.